torch/nn/utils/prune.py

```diff
@@ -0,0 +1,1269 @@
+            # the previous one.
+            method = cls(*args, **kwargs)  # new pruning
+            # Have the pruning method remember what tensor it's been applied to
+            method._tensor_name = name
+
+            # combine `methods` with `old_method`, if `old_method` exists
+            if old_method is not None:  # meaning that there was a hook
+                # if the hook is already a pruning container, just add the
+                # new pruning method to the container
+                if isinstance(old_method, PruningContainer):
+                    old_method.add_pruning_method(method)
+                    method = old_method  # rename old_method --> method
+
+                # if the hook is simply a single pruning method, create a
+                # container, add the old pruning method and the new one
+                elif isinstance(old_method, BasePruningMethod):
+                    container = PruningContainer(old_method)
+                    # Have the pruning method remember the name of its tensor
+                    # setattr(container, '_tensor_name', name)
+                    container.add_pruning_method(method)
+                    method = container  # rename container --> method
+            return method
+
+        method = _get_composite_method(cls, module, name, *args, **kwargs)
+        # at this point we have no forward_pre_hooks but we could have an
+        # active reparametrization of the tensor if another pruning method
+        # had been applied (in which case `method` would be a PruningContainer
+        # and not a simple pruning method).
+
+        # Apply pruning to the module
+        # original tensor, prior to this iteration of pruning (not like orig orig)
+        orig = getattr(module, name)
+
+        # If this is the first time pruning is applied, take care of moving 
+        # the original tensor to a new parameter called name + '_orig' and
+        # and deleting the original parameter
+        if not isinstance(method, PruningContainer):
+            # copy `module[name]` to `module[name + '_orig']`
+            module.register_parameter(name + "_orig", orig)
+            # temporarily delete `module[name]`
+            del module._parameters[name]
+            default_mask = torch.ones_like(orig)  # temp
+        # If this is not the first time pruning is applied, all of the above
+        # has been done before in a previos pruning iteration, so we're good
+        # to go
+        else:
+            default_mask = getattr(module, name + "_mask").detach().clone()
+
+        # Use try/except because if anything goes wrong with the mask 
+        # computation etc., you'd want to roll back.
+        try:
+            # get the final mask, computed according to the specific method
+            mask = method.compute_mask(orig, default_mask=default_mask)
+            # reparametrize by saving mask to `module[name + '_mask']`...
+            module.register_buffer(name + "_mask", mask)
```

Comment:
Ok, I think we should actually discuss this. There are several options to go about this, depending on what the user wants to do.
 
Let's start from the worst case. If you try to serialized a pruned model (that has parameters called `modulename.weight_orig`, buffers called `modulename.weight_mask`, and module attributes called `weight` -- assuming `name = 'weight'` in this example), if you use the simple `torch.save(pruned_model.state_dict(), PATH)`, then you can only reload this state dict into a similarly pruned model that has the same parameters and buffers. If you instead try to load this into an unpruned model, it won't work because of key mismatches (of course). 

Now, the point is that, if the intention is to reload it into an entirely new model that only shares the base architecture with the pruned model, then there are better ways to go about it. That's what we have `prune.remove` for. In other words, starting from a pruned model, you should first do `prune.remove(module, name)`, then save using `torch.save(pruned_model.state_dict(), PATH)`, and finally reload using `new_model.load_state_dict(torch.load(PATH))`. I think that's true for `weight_norm` as well.

Otherwise, simply save using `torch.save(pruned_model, PATH)`, and reload with `new_model = torch.load(PATH)`, which should be totally fine unless I'm ignoring some case.

Reply:
Anyways, I added two new tests to cover the two valid options discussed above.